### DISCUSSION TITLE
NTP config and restarting robot

### DISCUSSION THREAD

**Discussion by carlsondc-ceva** [iRobotEducation/create3_docs#324] [Questions] [answered]
TL;DR: what is the method of restarting the robot (not application) over its USB connection alluded to [here](https://iroboteducation.github.io/create3_docs/setup/compute-ntp/)? Is there a way to restrict the create 3 to only use its USB connection for NTP and not use public NTP servers?

I've noticed some issues that might be related to minor timing discrepancies between the Create 3 and the raspberry pi SBC I have running inside of it (operating an RPLidar and running the create3_lidar example). When replaying a bag file recorded on the SBC and rendering the result in RViz, I see intermittent "Lookup would require extrapolation into the future" errors-- discrepancies are on the order of 0.05 s. I see frequent occurrences of [this](https://github.com/iRobotEducation/create3_docs/discussions/116) from slam_toolbox, but I am not using namespaces (and I am getting /map output, etc), so I don't think that this is an issue of having an entirely missing set of transforms.

Per [these instructions](https://iroboteducation.github.io/create3_docs/setup/compute-ntp/), I set up chrony to serve time to the create3 over USB: 

I was able to configure chrony on the raspberry pi and confirm that the Create was attempting to use it per the output of `sudo chronyc clients`:
 
```
Hostname                      NTP   Drop Int IntL Last     Cmd   Drop Int  Last
===============================================================================
192.168.186.2                  28      0   8   -   379       0      0   -     -
```

However, when I look at the create3 logs, I see lines like:
```
Jan 30 19:00:49 create3-1 user.notice ntpd: ntpd: reply from 192.168.186.1: delay 0.010046 is too high, ignoring
Jan 30 19:00:53 create3-1 user.notice ntpd: ntpd: reply from 192.168.186.1: delay 0.009053 is too high, ignoring
Jan 30 19:01:06 create3-1 user.notice ntpd: ntpd: bad address '3.irobot.pool.ntp.org'
Jan 30 19:01:08 create3-1 user.notice ntpd: ntpd: timed out waiting for 192.168.186.3, reach 0x00, next query in 2s
Jan 30 19:12:26 create3-1 user.notice ntpd: ntpd: setting time to 2023-01-30 19:12:26.334381 (offset +677.150214s)
Jan 30 19:12:26 create3-1 user.notice ntpd: Time has been adjusted
```

So it looks like the Create 3 is reaching my Pi, but I'm not sure that the "Time has been adjusted" line is due to a response from the Pi or from a public NTP server. 

The last item in the NTP configuration instructions notes handling the "delay xxx is too high" as follows: "If this happens, simply restart the robot (not just the application) via the webserver over the USB network connection." I don't know if this applies in this situation, but if it does, **what is the command to restart the robot over the USB connection**? Since the SBC is powered from the USB connection, there is a bit of a chicken/egg situation here.

Alternately, **is there a way to make the create 3 ignore all other NTP servers and just use the USB?**

**Comment by shamlian** [comment, marked as answer]
Q for 1/2 of DR: What is the method of restarting the robot (not application) over its USB connection alluded to [here](https://iroboteducation.github.io/create3_docs/setup/compute-ntp/)?
A: When connected using Ethernet over USB, a link should be available in the "application" menu bar to restart the robot [(see here)](https://iroboteducation.github.io/create3_docs/webserver/application/#reboot-robot). There might be 😬 a bug in G.3 which causes that to not appear; in that case, please upgrade to [G.4.1](https://edu.irobot.com/create3/firmware/G.4.1). IIRC you can also use a tool like `curl` with something like `curl -X POST http://192.168.186.2/api/reboot`

Q for 2/2 of DR: Is there a way to restrict the create 3 to only use its USB connection for NTP and not use public NTP servers?
A: I don't believe so, but you _could_ take the Create 3 robot off Wi-Fi, and access the robot only via the SBC; if you are using Fast-DDS, you could run a Fast-DDS discovery server on the SBC and gain access to the robot's topics, actions, and services that way.

**Reply by carlsondc-ceva** [reply]
1/2: That `curl` command works perfectly! Very useful. Is this command (and any others) documented anywhere? I found the following by poking around:
- `curl -X POST http://192.168.186.2/api/reboot` 
- `curl -X POST http://192.168.186.2/api/restart-app` (which somewhat alarmingly responded with "Factory reset in progress.")
- `curl -X POST http://192.168.186.2/api/factory-reset`

I noticed that a lot of the functionality from the webserver uses WebSockets and so it's hard to e.g. read the "about" page without using a real web browser. Is this information accessible via a simpler method? Would be great to get diagnostics on the command line, especially when the robot is off wifi.

2/2: This seems to work pretty well- I can see stuff like this in the log:
```Feb  2 19:37:24 create3-1 user.notice ntpd: ntpd: bad address '0.irobot.pool.ntp.org'
Feb  2 19:37:24 create3-1 user.notice ntpd: ntpd: timed out waiting for 192.168.186.3, reach 0x00, next query in 2s
Feb  2 19:38:34 create3-1 user.notice ntpd: ntpd: setting time to 2023-02-02 19:38:33.821570 (offset +69.708135s)
Feb  2 19:38:34 create3-1 user.notice ntpd: Time has been adjusted
Feb  2 19:38:52 create3-1 user.notice ntpd: ntpd: bad address '3.irobot.pool.ntp.org'
```

Thanks!

**Reply by shamlian** [reply]
Thanks likewise for fuzz testing the webserver! ;-) One of these days, I'll get around to publishing the API... more to come.
